### PR TITLE
[Arch] Remove supports for Background Commands

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Dependency: https://github.com/OpenDevin/OpenDevin/pull/2802

`background_commands` has been existing for a while, but as far as I know, they are not used that often - it is less intuitive for an agent (at least for CodeAct) to specifically use the `background=True` argument.

Furthermore, the background commands can be simulated *without* actually supporting it in the sandbox: you can just run `python something.py > output.log &`. This will run that command on the backend and you get its output log whenever needed (e.g., by `cat output.log`).

The main reason for removing the background command is that:
1. It is probably less frequently used by agents / users
2. It introduces some specific logic for handling things (e.g., adding a field in the `State`, to handle it) - More code means harder to maintain, especially when we grow the number of runtime/sandboxes. Plus, this thing is actually broken for some boxes (e.g., E2B box).
3. And it gets in the way of our planned architecture refractor: https://github.com/OpenDevin/OpenDevin/issues/2404. As discussed with @yufansong , it is pretty challenging to make the observation of background commands synced into the event stream - how long should the background commands' output be sent to the event stream to update the agent? If we only send over the update when the agent requests it, why can't we simulate the same functionality with `python something.py > output.log &` and get the output using just `cat output.log`? It will behave the same way to the agent and reduce our maintenance headache.

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

This PR removes all occurrences of background commands as long as their dependencies, related prompts, etc.

**Other references**

### CODE REVIEW

**Comment by yufansong** [comment]
LGTM

**Review by yufansong — APPROVED** [review]

**Comment by enyst** [comment]
> Plus, this thing is actually broken for some boxes (e.g., E2B box).

Ah, I was wondering how the removal affects E2B. So we know it doesn't actually work now?

> Furthermore, the background commands can be simulated without actually supporting it in the sandbox: you can just run python something.py > output.log &. This will run that command on the backend and you get its output log whenever needed (e.g., by cat output.log).

Afaik, we don't have integration tests capturing background commands, but I think there is a bit in DummyAgent: it tests that two background commands consecutively works (with no action in-between iirc). So that case is completely removed? We expect the agent to always ask, so always an agent action to do `cat output.log`, that is, there will always be an agent action followed by observing the output?

File: opendevin/runtime/e2b/sandbox.py [hunk 1]
```diff
@@ -65,14 +61,6 @@ def _archive(self, host_src: str, recursive: bool = False):
                 tar.add(host_src, arcname=srcname)
         return tar_filename
 
-    # TODO: This won't work if we didn't wait for the background process to finish
```
Comment (xingyaoww) [inline opendevin/runtime/e2b/sandbox.py:68]: Re E2B: @enyst I think it doesn't really work _that_ well.

Comment (enyst) [reply opendevin/runtime/e2b/sandbox.py:68]: That's fair! 

Cc: @mlejva Hopefully this doesn't break anything, and it simplifies things.

**Comment by enyst** [comment]
> it is pretty challenging to make the observation of background commands synced into the event stream - how long should the background commands' output be sent to the event stream to update the agent? If we only send over the update when the agent requests it, why can't we simulate the same functionality with `python something.py > output.log &` and get the output using just `cat output.log`? It will behave the same way to the agent and reduce our maintenance headache.

Afaik the assumption of the current code is that they may be sent at regular intervals. That doesn't seem problematic from the perspective of the event stream updates, it will just notify when it has a new observation?

Regarding your second question though, it gets a bit more interesting IMHO. There is a difference between the two: whether we know or can identify easily in the backend what the agent is doing, and thus take some action on it to correct course or otherwise react to reduce its space of possibilities. The easiest example is the stuck functionality: in general, if we identify it's doing the same action over and over then we can act to stop it, at least. It matters for user experience, it really comes across as broken behavior when it keeps at it over and over uselessly. (of course, it would be even more interesting if we can make it correct course)

FWIW, when we first implemented the stuck functionality, we targeted only very specific patterns of behavior of the agent, e.g. when it gets errors, or continuous NullObservations, as opposed to the generic criterium "stop any 3 or 4 identical pairs". The reasons for that are in part simply that these patterns of behavior were so obvious with monologue agent, that we could target them specifically. It did work for a time, to my knowledge. Some of it even worked with CodeAct for a time or in part, until some feature (user messages come to mind) needed adaptation. But eventually, with CodeAct we have fallen back to a generic 'is this the exact same action, obs'.

A reason for that is that we have to, because the backend doesn't know enough about what is going on. Please see this example: ErrorObservation. [This section of is-stuck code](https://github.com/OpenDevin/OpenDevin/blob/9ccc64fa7eda6a9626afce88aaaca4957bd79e32/opendevin/controller/agent_controller.py#L422) never worked for CodeAct. Because the backend doesn't know when it gets an error, it doesn't get notified via an ErrorObservation. That functionality exists for monologue-based agents, and it does more than just stuck (it also has some self-correct mechanism related to it), but for CodeAct we never get that information. (Though, when linting is enabled, it might be easy to get it)

I wonder if something similar may happen with KillAction, which this PR is removing. The KillAction is simpler, and it gives us the possibility to identify as such (as a kill this pid) in more cases than when there's a more generic CmdOutputObservation, because with the CmdOutputObservation, the LLM varies slightly those thoughts, so oops, they're not the same anymore.

On a side note, maybe the problem is that the check is for identity. I've been thinking that perhaps we should try to identify by high similarity, rather. Maybe we can try, but there's always a trade-off, we don't want it to capture too much either, legitimate cases where the LLM could get out on its own.

Anyway, these are just some thoughts. I don't think they should block this removal. I do think it's worth thinking about sending some more information to the backend about what the agent is doing, and it's not only for stopping bad things, I think it may be necessary to allow us to slightly steer an agent at times.

**Comment by xingyaoww** [comment]
Thanks @enyst for your thoughtful comments!
Yeah i do think there are trade-off between more structured action, vs. pure command (e.g., `cat output.log`) which can be easy to support but very hard get structure information from for different purpose (e.g., stuck detection).

That said, I do think once we finish the refractor, we can add back these background commands if wanted. It is just that they will be implemented by something bash utility under the hood, similar to `python something.py &` and `cat output.log`. But we can then wrap the `CmdRunAction/Observation` around with some special action/observation (e.g., `BackgroundCmdObservation`) if we want that structure info.

For `ErrorObservation`, i do think it might be possible to get them from CodeAct, we can maybe turn a CmdRunObservation to ErrorObservation if the exit code of the command run is -1?

**Comment by xingyaoww** [comment]
(after battling with integration tests for a day - finally merging it 😢 )

File: opendevin/runtime/e2b/sandbox.py (hunk 1 above)
Comment (mlejva) [reply opendevin/runtime/e2b/sandbox.py:68]: Thanks for the pink @enyst . It should be good and it should simplify a lot of things I think. We have a new SDK update coming that will make things like background commands much easier if needed (the SDK is also much faster). 

It's already in beta release but we'll announce it on our socials ([Twitter](https://twitter.com/e2b_dev)) when prod ready.

**Pull request merged**